src/composables/useLoop/index.ts

```diff
@@ -0,0 +1,64 @@
+import type { EventDispatcher, Raycaster, WebGLRenderer } from 'three'
+import { unref } from 'vue'
+import type { TresCamera, TresScene } from '../../types'
+import { useTresContext } from '../useTresContextProvider'
+import type { LoopCallback } from '../../core/loop'
+import type { SizesType } from '../useSizes'
+
+export interface LoopParams extends LoopCallback {
+  camera: TresCamera
+  scene: TresScene
+  renderer: WebGLRenderer
+  invalidate: () => void
+  advance: () => void
+  raycaster: Raycaster
+  controls: (EventDispatcher & { enabled: boolean })
+  sizes: SizesType
+}
+
+export type Fn = (params: LoopParams) => void
+
+export function useLoop() {
+  const {
+    camera,
+    scene,
+    renderer,
+    loop,
+    raycaster,
+    controls,
+    invalidate,
+    advance,
+  } = useTresContext()
+
+  function onBeforeRender(cb: Fn, index = 0) {
+    const wrappedCallback = (params: LoopCallback) => {
+      cb({ ...params, camera: unref(camera) as TresCamera, scene: unref(scene), renderer: unref(renderer), raycaster: unref(raycaster), controls: unref(controls), invalidate, advance })
+    }
+    return loop.register(wrappedCallback, 'before', index)
```

Comment:
I think there is a general flaw with the structure of how the callbacks are registered. References to the callbacks should not survive if a component that registered that particular callback is unmounted. Due to lack of time, I haven't checked, but I think this is currently the case.
@alvarosabu If you need help solving this, let me know. I think I have a solution in mind.

Reply:
I just tested it and the hooks seems to be unregister when the component is `unmounted`

The sphere has a v-if, the sphere has several hooks registered, when is gone the console.logs of those hooks are gone too.

Dunno if it's something that `createEventHook` does under the hood, same for @andretchen0 solution

https://github.com/Tresjs/tres/assets/4699008/c82b2661-a9c1-43a8-8dc9-1984400926b5

Reply:
@Tinoooo @alvarosabu 

> I think there is a general flaw with the structure of how the callbacks are registered. References to the callbacks should not survive if a component that registered that particular callback is unmounted. 

Like `useEventHook`, `usePriorityEventHook` can mostly outsource this to Vue. 

Here's demo code if you'd like to verify it covers the case you're thinking about @Tinoooo . `tick` is not `console.log`ged when the component is not `v-if`d.

```html
<script setup lang="ts">
import { TresCanvas } from '@tresjs/core'
import Component from './Component.vue'
import { shallowRef } from 'vue'

const isVisible = shallowRef(true)
const animate = () => {
  isVisible.value = !!(Math.floor(Date.now() * 0.001) % 2)
  requestAnimationFrame(animate)
}

requestAnimationFrame(animate)

</script>

<template>
  <TresCanvas>
    <Component v-if="isVisible" />
  </TresCanvas>
</template>
```

```html
<script setup lang="ts">
import { useLoop } from '@tresjs/core'

useLoop().onBeforeRender(({ elapsed }) => {
  console.log('tick')
})
</script>

<template>
    <TresMesh>
      <TresBoxGeometry />
      <TresMeshNormalMaterial />
    </TresMesh>
</template>
```

----

> Hey, @Tinoooo so if I understand correctly then the end user will need to off when onBeforeUnmount right?

If we did things right, that shouldn't be necessary in regular components. It will (probably?) be necessary in things like directives if those end up using `useLoop`.

----

I should mention that I haven't looked into the guts of @alvarosabu 's implementation of `useLoop` to make sure that callbacks are removed for unmounted components – rather than simply not being called. That would be a good thing to check.

----

It would probably be best to avoid wrapping the callbacks, if possible.

Reply:
> Dunno if it's something that createEventHook does under the hood, same for @andretchen0 solution

Here's the magic:

https://github.com/Tresjs/tres/blob/d547ab159b09eb46ad5e074c63701c3035cec12c/src/utils/createPriorityEventHook.ts#L52

Reply:
> I should mention that I haven't looked into the guts of @alvarosabu 's implementation of useLoop to make sure that callbacks are removed for unmounted components – rather than simply not being called. That would be a good thing to check.

@andretchen0 good news, I double-checked the same example as the video with a breakpoint on the loop to check the `count` property and when the sphere is unmounted it goes from `3` -> `1`, which is correct because the sphere registers 2 callbacks. So I can confirm it not only doesn't call them but they are also gone.


![Screenshot 2024-05-07 at 19 24 06](https://github.com/Tresjs/tres/assets/4699008/71b4f819-3a22-4e1e-b665-d76ce2ebc57d)


I also checked if the sphere is mounted again after, and there are no duplicates, the count goes to 3 again which is expected